djangoplicity/archives/base.py

```diff
@@ -818,3 +828,87 @@ def get_object_identifier_for_pk( cls, pk='' ):
         return str( pk )
         #return "%s.%s:%s" % ( self._meta.app_label, self._meta.model_name, str( self.pk ) )
+    
+    def get_access_tag(self):
+        """
+        Return the access tag for this object to sync with S3
+        Rules:
+        - No Published: Private
+        - current_date < release_date < embargo_date: Private
+        - current_date < embargo_date < release_date: Private
+        - release_date < current_date < embargo_date: Private
+        - embargo_date < current_date < release_date: Private
+        - release_date < embargo_date < current_date: Public
+        - embargo_date < release_date < current_date: Public
+        """
+        now = datetime.now()
+
+        # Only Admins can access
+        if hasattr(self, 'published') and not self.published:
+            return 'Private'
+        
+        # If the object has no release_date or embargo_date, it is public
+        if not hasattr(self, 'release_date') or not hasattr(self, 'embargo_date'):
+            return 'Public'
+
+        embargo_date = getattr(self, 'embargo_date', None)
+        release_date = getattr(self, 'release_date', None)
+
+        if not embargo_date or not release_date:
+            return 'Public'
+
+        if now < release_date < embargo_date:
+            return 'Private'
+        elif now < embargo_date < release_date:
+            return 'Private'
+        elif release_date < now < embargo_date:
+            return 'Private'
+        elif embargo_date < now < release_date:
+            return 'Private'
+        elif release_date < embargo_date < now:
+            return 'Public'
+        elif embargo_date < release_date < now:
+            return 'Public'
+        else:
+            return 'Public'
+    
+    def is_embargoed(self):
+        embargo_date = getattr(self, 'embargo_date', None)
+        release_date = getattr(self, 'release_date', None)  
+
+        if not embargo_date and not release_date:
+            return False
+
+        now = datetime.now()
+        if embargo_date < now < release_date:
+            return True
+        return False
+    
+    def get_access_tag_for_format(self, fmt):
+        base_access_tag = self.get_access_tag()
+        is_embargoed = self.is_embargoed()
+        always_public = getattr(settings, 'S3_ALWAYS_PUBLIC_FORMATS', [])
+
+        if is_embargoed and fmt in always_public:
+            return 'Public'
+
+        return base_access_tag
+
+    def _update_access_tag_with_s3_server(self):
+        """
+        Update the access tag for this object to sync with S3
+        """
+        if not getattr(settings, 'USE_PROXY_FOR_PRIVATE_MEDIA', False):
+            return
+
+        if not hasattr(self, 'content_server') or not self.content_server:
+            return
+        
+        content_server = MEDIA_CONTENT_SERVERS[self.content_server]
+        if not content_server:
+            return
+        
+        if not isinstance(content_server, S3ContentServer):
+            return
```

Comment:
Validate instead that the content server has protection capabilities, or just validate the existence of the function